server/src/main/java/org/opensearch/index/ShardIndexingPressureStore.java

```diff
@@ -0,0 +1,121 @@
+ * trackers its can hold at any given time, which is a configurable dynamic setting.
+ *
+ * Tracking objects when created are part of both the hot store as well as cold store. However, once the object
+ * is no more live it is removed from the hot store. Objects in the cold store are evicted once the cold store
+ * reaches its maximum limit. Think of it like a periodic purge when upper limit is hit.
+ * During get if tracking object is not present in the hot store, a lookup is made into the cache store. If found,
+ * object is brought into the hot store again, until it remains active. If not present in the either store, a fresh
+ * object is instantiated and registered in both the stores for concurrent accesses.
+ *
+ * Note: The implementation of shardIndexingPressureColdStore methods is such that get,
+ * update and evict operations can be abstracted out to support any other strategy such as LRU, if
+ * discovered a need later.
+ *
+ */
+public class ShardIndexingPressureStore {
+
+    // This represents the maximum value for the cold store size.
+    public static final Setting<Integer> MAX_COLD_STORE_SIZE =
+        Setting.intSetting("shard_indexing_pressure.cache_store.max_size", 200, 100, 1000,
+            Setting.Property.NodeScope, Setting.Property.Dynamic);
+
+    private final Map<ShardId, ShardIndexingPressureTracker> shardIndexingPressureHotStore =
+        ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency();
+    private final Map<ShardId, ShardIndexingPressureTracker> shardIndexingPressureColdStore = new HashMap<>();
+    private final ShardIndexingPressureSettings shardIndexingPressureSettings;
+
+    private volatile int maxColdStoreSize;
+
+    public ShardIndexingPressureStore(ShardIndexingPressureSettings shardIndexingPressureSettings,
+                                      ClusterSettings clusterSettings, Settings settings) {
+        this.shardIndexingPressureSettings = shardIndexingPressureSettings;
+        this.maxColdStoreSize = MAX_COLD_STORE_SIZE.get(settings).intValue();
+        clusterSettings.addSettingsUpdateConsumer(MAX_COLD_STORE_SIZE, this::setMaxColdStoreSize);
+    }
+
+    public ShardIndexingPressureTracker getShardIndexingPressureTracker(ShardId shardId) {
+        ShardIndexingPressureTracker tracker = shardIndexingPressureHotStore.get(shardId);
+        if (isNull(tracker)) {
+            // Attempt from Indexing pressure cold store
+            tracker = shardIndexingPressureColdStore.get(shardId);
+            // If not already present in cold store instantiate a new one
+            if (isNull(tracker)) {
+                ShardIndexingPressureTracker newShardIndexingPressureTracker = new ShardIndexingPressureTracker(shardId,
+                    this.shardIndexingPressureSettings.getShardPrimaryAndCoordinatingBaseLimits(),
+                    this.shardIndexingPressureSettings.getShardReplicaBaseLimits());
```

Comment:
Here we would be doing unnecessary allocation if the tracker already exists. Suggest 
```suggestion
                tracker = shardIndexingPressureHotStore.computeIfAbsent(shardId, (k) -> {
                    ShardIndexingPressureTracker newTracker = new ShardIndexingPressureTracker(shardId, this.shardIndexingPressureSettings.getShardPrimaryAndCoordinatingBaseLimits(), this.shardIndexingPressureSettings.getShardReplicaBaseLimits());
                    return newTracker;
                });
```

Reply:
Yes. Tradeoff is that the attempted update operations on this map by other threads will be blocked during compute. But since this computation should be short and simple, it makes sense.